github-actions: Add bionic-frontend-backend job.

priyank-p opened the pull request:
**Testing Plan:** GitHub Actions CI passes successfully.

timabbott commented on tools/lib/provision.py:
```diff
@@ -37,7 +37,7 @@
 VAR_DIR_PATH = os.path.join(ZULIP_PATH, 'var')
 
-is_circleci = 'CIRCLECI' in os.environ
+is_ci = 'CI' in os.environ
```
Let's call this `CONTINUOUS_INTEGRATION`; `CI` isn't unique enough of a name.

timabbott replied:
(On the LHS; RHS presumably needs a more real check)

andersk replied:
We don’t really get a choice:

https://circleci.com/docs/2.0/env-vars/#built-in-environment-variables
https://help.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables

We could test for `os.environ.get("CI") == "true"` to be slightly more specific.

timabbott replied:
Well, we could do an `||` of the values for two options.  But I suppose that check is probably OK.

priyank-p replied:
I guess being specific never hurts so I went with `'CIRCLECI' in os.environ or 'GITHUB_ACTIONS' in os.environ`.

timabbott commented on .gitignore:
```diff
@@ -32,7 +32,7 @@ package-lock.json
 
 # Dockerfiles generated for CircleCI
-/tools/circleci/images
+/tools/github_actions/images
```
Maybe these should just move into the existing `tools/ci` instead?

timabbott commented on tools/github_actions/Dockerfile.template:
```diff
@@ -96,11 +96,11 @@ RUN apt-get update \
   && echo 'Defaults    env_keep += "DEBIAN_FRONTEND"' >> /etc/sudoers.d/env_keep
 
-USER circleci
+USER github
```
We should username a variable at the top of the file so it only needs to change value once?

andersk commented:
While we’re adding variables to the Dockerfile, we could also replace the silly `generate-dockerfiles` template mechanism with variables too, if you want.

priyank-p commented:
@andersk to be clear, you mean use `ARG` for `base_image` and `extra_packages` so we can do `sudo docker build <path> --build-arg base_image=...`

priyank-p commented:
I am trying to figure out why using cache from bionic causes error in focal like we do in circleci. So we should hold off on merging last two commits. I'd like to know if the approach in `provision: Do service <service-name> start instead of restart.` is plausible to avoid flakes in Focal.

This is the relevant step that causes error:
```
        # TODO: Doing this on Focal causes errors later in provision
        # possibly because we are reusing cache:
        #   Could not find platform independent libraries <prefix>
        #   Could not find platform dependent libraries <exec_prefix>
        #   Consider setting $PYTHONHOME to <prefix>[:<exec_prefix>]
        #   Fatal Python error: Py_Initialize: Unable to get the locale encoding
        #   ModuleNotFoundError: No module named 'encodings'
        #   Current thread 0x00007fb7efa04740 (most recent call first):
        #   Provisioning failed (exit code 134)!
        # sudo chmod -R 0777 /__w/_temp/
```

And, CircleCI Builds: https://circleci.com/gh/priyank-p/zulip/tree/github-actions

timabbott commented on tools/lib/provision.py:
```diff
@@ -399,10 +399,10 @@ def main(options: argparse.Namespace) -> "NoReturn":
+        run_as_root(["service", "redis-server", "start"])
+        run_as_root(["service", "memcached", "start"])
+        run_as_root(["service", "rabbitmq-server", "start"])
+        run_as_root(["service", "postgresql", "start"])
```
Yeah, I think I agree it's reasonable to just start here; and we should know in the CI environment startup whether the services are already running.  I'd check that none of these fail if run twice; we do rerun provision and parts thereof to deal with network failures at times in CI.

priyank-p replied:
There are no errors when running these twice in CI runs. See [this run](https://github.com/priyank-p/zulip/actions/runs/151727880) based on [this commit](https://github.com/priyank-p/zulip/commit/2c5a763cdea20cfbf6022debf0d891e49f91102b).

priyank-p commented:
The issue was focal build was that we were using cache from bionic build. I missed the part that adds the job name as a cache restore key when migrating this from CircleCI config. Both bionic and focal jobs run successfully now.

timabbott commented:
I merged the first 3 commits.  @arpit551 @amanagr can you help review the new GitHub Actions configuration?  I feel like someone should at least carefully compare it with the old CircleCI logic and make sure we have comments around any changes.

I'd also be on board with making `.template.yml` using YAML anchors and having a tool that reads it in and writes the result back out for GitHub's use to let us reuse code more effectively.

arpit551 commented on .github/workflows/zulip-ci.yml:
```diff
@@ -0,0 +1,157 @@
+      uses: actions/cache@v2
+      with:
+        path: /srv/zulip-emoji-cache
+        key: v1-emoji-${{ github.job }}-${{ hashFiles('requirements/thumbor-dev.txt') }}-${{ hashFiles('requirements/dev.txt') }}
```
I think these are not the keys we need for the emoji cache. The one we use in Circle CI config should be here.

priyank-p replied:
Ahh, this is a copy-paste error which I think I missed because we `v1-venv` twice. Fixed, thanks.

arpit551 commented:
This looks good regarding the code reusability thing I am not sure it will be straight forward. There are some issues regarding it [https://github.community/t/support-for-yaml-anchors/16128/33](https://github.community/t/support-for-yaml-anchors/16128/33),
[https://github.community/t/reusing-sharing-inheriting-steps-between-jobs-declarations/16851/3](https://github.community/t/reusing-sharing-inheriting-steps-between-jobs-declarations/16851/3).

andersk commented:
Can we not use [`strategy`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-running-with-more-than-one-operating-system) to run the same steps on multiple platforms, along with variables and [`if` conditionals](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsif) to handle differences between platforms?

priyank-p commented:
Nice, that seems like a interesting alternative.

priyank-p commented:
I added two more commits for minor cleanups. For the `strategy` commit I used this script to verify all the steps are correctly migrated.
<details>
  <summary>Script</summary>
  
```python
"""
This script prints out the ignore steps first. Then
prints out each step of both circle and actions side-by-side.
One step is out of order for bionic but verfying correction is
still easier. Format:
Actions: Install dependencies
Circle CI: install dependencies
...
"""
import yaml

with open('.circleci/config.yml') as f:
    circleci_config = yaml.safe_load(f)

with open('.github/workflows/zulip-ci.yml') as f:
    actions_config = yaml.safe_load(f)

circle_bionic_steps = []
circle_focal_steps = []
actions_bionic_steps = []
actions_focal_steps = []

# We ignore casper artifact upload, save_cache, and
# store_tests_reports steps.
def get_circleci_steps(job, arr):
    for step in circleci_config['jobs'][job]['steps']:
        if isinstance(step, str):
            arr.append(step)
            continue

        step_name = step.get('run', {}).get('name', False)
        if not step_name:
            if step.get('restore_cache'):
                key = step['restore_cache']['keys'][0].split('.')[0]
                step_name = f'<restore-cache> {key}'
            elif step.get('store_artifacts', False):
                destination = step['store_artifacts']['destination']
                step_name = f'<store-artificats> {destination}'

                if destination == 'casper':
                    # This is no longer needed
                    print('Ignoring step:')
                    print(step)
                    print()
                    continue
            else:
                # We don't care about save_cache; github-actions
                # does this automatically, and store_tests_reports
                # is circelci timing specific.
		print('Ignoring step:')
                print(step)
                print()
                continue

        if step_name != 'On fail':
            arr.append(step_name)

get_circleci_steps('bionic-backend-frontend', circle_bionic_steps)
get_circleci_steps('focal-backend', circle_focal_steps)

# We ignore there steps specific to github-actions
for step in actions_config['jobs']['focal_bionic']['steps']:
    BOTH_OS = 'BOTH_OS'
    if_check = step.get('if', BOTH_OS)
    step_name = step.get('name')

    if step_name is None:
        step_name = step['uses']

    if (
        step_name == 'Upgrade git for bionic' or
        step_name == 'Add required permissions' or
        step_name == 'Move test reports to var'
    ):
        print('Ignoring step:')
        print(step)
        print()
        # These are github-actions specific; see comments in config
        continue

    if if_check == BOTH_OS:
        actions_bionic_steps.append(step_name)
        actions_focal_steps.append(step_name)
    elif 'is_bionic' in if_check:
        actions_bionic_steps.append(step_name)
    else:
        actions_focal_steps.append(step_name)

bionic = zip(circle_bionic_steps, actions_bionic_steps)
focal = zip(circle_focal_steps, actions_focal_steps)

print('Bionic steps:')
for (circle_step, actions_step) in bionic:
    print(f'CircleCI: {circle_step}')
    print(f'Actions: {actions_step}')
    print()

print('Focal steps:')
for (circle_step, actions_step) in focal:
    print(f'CircleCI: {circle_step}')
    print(f'Actions: {actions_step}')
    print()
```
</details>

timabbott commented:
This seems good enough that it's worth merging and we can iterate from here.  Huge thanks for prototyping this @priyank-p!

I think we might want to expand the flags we set on `matrix` to be more semantic (E.g. "include_frontend_tests" for the conditional for things that are about that, not about the OS version).

Please post in #backend with updates on this, and we should start working on moving the production suites.
